File: 034_search_for_a_range/binary_search.py
class Solution:
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        """
        Complexity: O(logn)
        Space: O(1)
        """
        
        def searchRange_sub(nums, start, end, target):
            if start == end:
                if nums[start] is not target:
                    return [-1,-1]
                else:
                    return [start, end]
            elif start > end:
                return [-1, -1]
            mid = nums[(end+start) // 2] 
            if mid > target:
                return searchRange_sub(nums, start, (end+start)//2 - 1,  target)
            elif mid < target:
                return searchRange_sub(nums, (end+start)//2 + 1, end, target)
            else:
                left = searchRange_sub(nums, start, (end+start)//2 - 1, target)
                right = searchRange_sub(nums, (end+start)//2 + 1, end, target)
                ans = []
                if left[0] is not -1 :
                    ans.append(left[0])
                else:
                    ans.append((end+start)//2)
                if right[1] is not -1 :
                    ans.append(right[1])
                else:
                    ans.append((end+start)//2)
                return ans
        
        n = len(nums)
        return searchRange_sub(nums, 0, n-1, target)

File: 034_search_for_a_range/binary_search.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int n = nums.size();
        if(!n){
            return vector<int>(2, -1);
        }
        int start = 0, end = n-1, right, left;
        // find the start of the range
        while(start <= end){
            int mid = (start + end) / 2;
            if(nums[mid] >= target) end = mid - 1;
            // if mid < target, then we start from the next pos. that is >=
            // target
            // if there is no target than the left would be the first element 
            // bigger than the target
            else start = mid + 1;
        }
        left = start;
        end = n - 1;
        while(start <= end){
            int mid = (start + end) / 2;
            if(nums[mid] > target) end = mid - 1;
            // if mid <= target, we start from the next pos that would be bigger
            // than the target
            // if there is no target than this condition would never be meet
            // this left would be equal to right
            else start = mid + 1;
        }
        right = start;
        return (left == right ? vector<int>{-1,-1} : vector<int>{left, right - 1});
    }
};
